examples/src/bin/unbuffered-server.rs

```diff
@@ -0,0 +1,255 @@
+                    };
+
+                    let http_response = build_http_response();
+                    try_or_resize_and_retry(
+                        |out_buffer| state.encrypt(http_response, out_buffer),
+                        map_err,
+                        outgoing_tls,
+                        &mut outgoing_used,
+                    )?;
+                    sent_response = true;
+
+                    try_or_resize_and_retry(
+                        |out_buffer| state.queue_close_notify(out_buffer),
+                        map_err,
+                        outgoing_tls,
+                        &mut outgoing_used,
+                    )?;
+                    open_connection = false;
+
+                    send_tls(&mut sock, outgoing_tls, &mut outgoing_used)?;
+                }
+            }
+
+            _ => unreachable!(),
+        }
+
+        if discard != 0 {
+            assert!(discard <= incoming_used);
+
+            incoming_tls.copy_within(discard..incoming_used, 0);
+            incoming_used -= discard;
+
+            eprintln!("discarded {discard}B from `incoming_tls`");
+        }
+
+        iter_count += 1;
+        assert!(
+            iter_count < MAX_ITERATIONS,
+            "did not get a HTTP response within {MAX_ITERATIONS} iterations"
+        );
+    }
+
+    assert!(received_request);
+    assert!(sent_response);
+    assert_eq!(0, incoming_used);
+    assert_eq!(0, outgoing_used);
+
+    Ok(())
+}
+
+fn try_or_resize_and_retry<E>(
+    mut f: impl FnMut(&mut [u8]) -> Result<usize, E>,
+    map_err: impl FnOnce(E) -> Result<InsufficientSizeError, Box<dyn Error>>,
+    outgoing_tls: &mut Vec<u8>,
+    outgoing_used: &mut usize,
+) -> Result<usize, Box<dyn Error>>
+where
+    E: Error + 'static,
+{
+    let written = match f(&mut outgoing_tls[*outgoing_used..]) {
+        Ok(written) => written,
+
+        Err(e) => {
+            let InsufficientSizeError { required_size } = map_err(e)?;
+            let new_len = *outgoing_used + required_size;
+            outgoing_tls.resize(new_len, 0);
+            eprintln!("resized `outgoing_tls` buffer to {new_len}B");
+
+            f(&mut outgoing_tls[*outgoing_used..])?
+        }
+    };
+
+    *outgoing_used += written;
+
+    Ok(written)
+}
+fn recv_tls(
+    sock: &mut TcpStream,
+    incoming_tls: &mut [u8],
+    incoming_used: &mut usize,
+) -> Result<(), Box<dyn Error>> {
+    let read = sock.read(&mut incoming_tls[*incoming_used..])?;
+    eprintln!("received {read}B of data");
+    *incoming_used += read;
+    Ok(())
+}
+
+fn send_tls(
+    sock: &mut TcpStream,
+    outgoing_tls: &[u8],
+    outgoing_used: &mut usize,
+) -> Result<(), Box<dyn Error>> {
+    sock.write_all(&outgoing_tls[..*outgoing_used])?;
+    eprintln!("sent {outgoing_used}B of data");
+    *outgoing_used = 0;
+    Ok(())
+}
+
+fn build_http_response() -> &'static [u8] {
+    b"HTTP/1.0 200 OK\r\nConnection: close\r\n\r\nHello world from rustls unbuffered server\r\n"
+}
```

Comment:
same feedback as in the client example that this fn might not be useful enough on its own to justify being separated from the call-site